source/custom/nidaqmx_service.custom.cpp

```diff
@@ -10,4 +101,139 @@ ::grpc::Status NiDAQmxService::ConvertApiErrorStatusForTaskHandle(::grpc::Server
   return nidevice_grpc::ApiErrorAndDescriptionToStatus(context, status, description);
 }
 
+::grpc::Status NiDAQmxService::ReadAnalogWaveforms(::grpc::ServerContext* context, const ReadAnalogWaveformsRequest* request, ReadAnalogWaveformsResponse* response)
+{
+  if (context->IsCancelled()) {
+    return ::grpc::Status::CANCELLED;
+  }
+  try {
+    auto task_grpc_session = request->task();
+    TaskHandle task = session_repository_->access_session(task_grpc_session.name());
+
+    const auto number_of_samples_per_channel = request->number_of_samples_per_channel();
+    const auto timeout = request->timeout();
+    
+    int32 waveform_attribute_mode;
+    switch (request->waveform_attribute_mode_enum_case()) {
+      case nidaqmx_grpc::ReadAnalogWaveformsRequest::WaveformAttributeModeEnumCase::kWaveformAttributeMode: {
+        waveform_attribute_mode = static_cast<int32>(request->waveform_attribute_mode());
+        break;
+      }
+      case nidaqmx_grpc::ReadAnalogWaveformsRequest::WaveformAttributeModeEnumCase::kWaveformAttributeModeRaw: {
+        waveform_attribute_mode = static_cast<int32>(request->waveform_attribute_mode_raw());
+        break;
+      }
+      case nidaqmx_grpc::ReadAnalogWaveformsRequest::WaveformAttributeModeEnumCase::WAVEFORM_ATTRIBUTE_MODE_ENUM_NOT_SET: {
+        return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The value for waveform_attribute_mode was not specified or out of range");
+        break;
+      }
+    }
+
+    uInt32 num_channels = 0;
+    auto status = library_->GetReadAttributeUInt32(task, ReadUInt32Attribute::READ_ATTRIBUTE_NUM_CHANS, &num_channels);
+    if (!status_ok(status)) {
+      return ConvertApiErrorStatusForTaskHandle(context, status, task);
+    }
+
+    if (num_channels == 0) {
+      return ::grpc::Status(::grpc::INVALID_ARGUMENT, "No channels found in task");
```

Comment:
READ_ATTRIBUTE_NUM_CHANS isn't the number of channels in the task, it's the number of channels specified by DAQmx_Read_ChannelsToRead.